File: node_ffi_example/factorial.js
var ffi = require('./node-ffi/')
var refStruct = require('ref-struct');
var refArray = require('ref-array');
var ref = require('ref');

//gps date struct
var gps = refStruct({
	'lon':ref.types.double,
	'lat':ref.types.double,
});

var gpsStructArrayType = refArray(gps);
var gps_info = gpsStructArrayType(1);

//user date struct
var user_info_t = refStruct({
	'user_name':'string',
	'passwd':'string',
	'dev_id':'string',
});

var userStructArrayType = refArray(user_info_t);
var user_info = userStructArrayType(1);


//define c rountine
var libfactorial = ffi.Library('./libfactorial', {
  'get_gps_info': [ gpsStructArrayType, [ 'void' ] ],
  'put_user_info': [ 'void', [ userStructArrayType ] ],
});


//get gps info from c process
gps_info = libfactorial.get_gps_info(0)
console.log('get gps lon: ' + gps_info[0].lon)
console.log('get gps lat: ' + gps_info[0].lat)

//put user info to c process
user_info[0].user_name = 'zdd';
user_info[0].passwd = '123';
user_info[0].dev_id = '65535';
libfactorial.put_user_info(user_info)
console.log('put user_name: ' + user_info[0].user_name)
console.log('put passwd: ' + user_info[0].passwd)
console.log('put dev_id: ' + user_info[0].dev_id)

module.exports.gps_info = gps_info;
module.exports.user_info = user_info;
module.exports.libfactorial=libfactorial;

File: node_ffi_example/factorial.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <errno.h>

#if defined(WIN32) || defined(_WIN32)
#define EXPORT __declspec(dllexport)
#else
#define EXPORT
#endif

//gps info 
typedef struct
{
	double lon;
	double lat;
}gps_info_t;
gps_info_t* gps;

//user info
typedef struct
{
	char *user_name;
	char *passwd;
	char *dev_id;
}user_info_t;

EXPORT gps_info_t* get_gps_info(void) {
  gps = (gps_info_t*)malloc(sizeof(gps_info_t));//这里仅测试用，实际就放到初始化的地方去吧
  gps->lon+=1;
  return gps;
}

EXPORT void put_user_info(user_info_t* user_info) {
  printf("get userinfo(user_name):%s \n",user_info->user_name);
  printf("get userinfo(user_name):%s \n",user_info->passwd);
  printf("get userinfo(user_name):%s \n",user_info->dev_id);

}



